File: app/src/transactions/application/MovementVoiceEntry/interpretationResultToMovementEntryDraftMapper.ts
import type {
  InterpretationFieldCandidate,
  InterpretationFieldResult,
  InterpretationResult,
} from '../../../core/application/interpretation/schemaGuidedInterpretationContract';
import type {
  MovementEntryDraft,
  MovementEntryDraftField,
  MovementEntryDraftIssue,
  MovementEntryDraftIssueCode,
  MovementEntryDraftType,
} from './MovementEntryDraftInterpreterPort';
import {
  GONEZO_MOVEMENT_ENTRY_SPEC_ID,
  GONEZO_MOVEMENT_ENTRY_SPEC_VERSION,
  MOVEMENT_ENTRY_INTERPRETATION_FIELDS,
} from './movementEntryInterpretationContract';

type DraftFieldKey = typeof MOVEMENT_ENTRY_INTERPRETATION_FIELDS[keyof typeof MOVEMENT_ENTRY_INTERPRETATION_FIELDS];

const draftFields = new Set<DraftFieldKey>(Object.values(MOVEMENT_ENTRY_INTERPRETATION_FIELDS));

export class InterpretationResultToMovementEntryDraftMapper {
  private readonly minimumConfidence: number;

  constructor(minimumConfidence = 0.5) {
    if (minimumConfidence < 0 || minimumConfidence > 1) {
      throw new Error('minimum confidence must be between 0 and 1');
    }
    this.minimumConfidence = minimumConfidence;
  }

  map(result: InterpretationResult): MovementEntryDraft {
    if (result.specId !== GONEZO_MOVEMENT_ENTRY_SPEC_ID || result.specVersion !== GONEZO_MOVEMENT_ENTRY_SPEC_VERSION) {
      throw new Error('interpretation result schema does not match the movement entry contract');
    }

    const issues: MovementEntryDraftIssue[] = [];
    const fieldByKey = new Map(result.fields.map((field) => [field.key, field]));

    const type = this.mapType(fieldByKey.get(MOVEMENT_ENTRY_INTERPRETATION_FIELDS.type), issues);
    const amount = this.mapAmount(fieldByKey.get(MOVEMENT_ENTRY_INTERPRETATION_FIELDS.amount), issues);
    const occurredOn = this.mapDate(fieldByKey.get(MOVEMENT_ENTRY_INTERPRETATION_FIELDS.occurredOn), issues);
    const note = this.mapText(fieldByKey.get(MOVEMENT_ENTRY_INTERPRETATION_FIELDS.note), issues);
    const categoryId = this.mapCategory(fieldByKey.get(MOVEMENT_ENTRY_INTERPRETATION_FIELDS.categoryId), issues);

    for (const issue of result.issues) {
      issues.push({
        field: issue.fieldKey ? toDraftField(issue.fieldKey) : undefined,
        code: 'output_incompatible',
      });
    }

    return {
      type,
      amount,
      occurredOn,
      note,
      categoryId,
      issues,
    };
  }

  private mapType(field: InterpretationFieldResult | undefined, issues: MovementEntryDraftIssue[]): MovementEntryDraftType | undefined {
    const candidate = this.interpretationOf(field, 'type', issues, false);
    if (!candidate) {
      return undefined;
    }
    if (candidate.confidence < this.minimumConfidence) {
      issues.push(issue('type', 'confidence_insufficient'));
      return undefined;
    }
    if (candidate.value.type !== 'enum') {
      issues.push(issue('type', 'value_invalid'));
      return undefined;
    }
    if (candidate.value.value === 'expense') {
      return 'expense';
    }
    if (candidate.value.value === 'income') {
      return 'income';
    }
    issues.push(issue('type', 'value_invalid'));
    return undefined;
  }

  private mapAmount(field: InterpretationFieldResult | undefined, issues: MovementEntryDraftIssue[]): string | undefined {
    const candidate = this.interpretationOf(field, 'amount', issues, false);
    if (!candidate) {
      return undefined;
    }
    if (candidate.confidence < this.minimumConfidence) {
      issues.push(issue('amount', 'confidence_insufficient'));
      return undefined;
    }
    if (candidate.value.type !== 'decimal') {
      issues.push(issue('amount', 'value_invalid'));
      return undefined;
    }
    const amount = String(candidate.value.value).trim();
    if (!amount || Number(amount) <= 0) {
      issues.push(issue('amount', 'value_invalid'));
      return undefined;
    }
    return amount;
  }

  private mapDate(field: InterpretationFieldResult | undefined, issues: MovementEntryDraftIssue[]): string | undefined {
    const candidate = this.interpretationOf(field, 'occurredOn', issues, true);
    if (!candidate) {
      return undefined;
    }
    if (candidate.confidence < this.minimumConfidence) {
      issues.push(issue('occurredOn', 'confidence_insufficient'));
      return undefined;
    }
    if (candidate.value.type !== 'date') {
      issues.push(issue('occurredOn', 'value_invalid'));
      return undefined;
    }
    return String(candidate.value.value);
  }

  private mapText(field: InterpretationFieldResult | undefined, issues: MovementEntryDraftIssue[]): string | undefined {
    const candidate = this.interpretationOf(field, 'note', issues, false);
    if (!candidate) {
      return undefined;
    }
    if (candidate.confidence < this.minimumConfidence) {
      issues.push(issue('note', 'confidence_insufficient'));
      return undefined;
    }
    if (candidate.value.type !== 'text') {
      issues.push(issue('note', 'value_invalid'));
      return undefined;
    }
    const note = String(candidate.value.value).trim();
    return note || undefined;
  }

  private mapCategory(field: InterpretationFieldResult | undefined, issues: MovementEntryDraftIssue[]): string | undefined {
    const candidate = this.interpretationOf(field, 'categoryId', issues, false);
    if (!candidate) {
      return undefined;
    }
    if (candidate.confidence < this.minimumConfidence) {
      issues.push(issue('categoryId', 'confidence_insufficient'));
      return undefined;
    }
    if (candidate.value.type !== 'enum') {
      issues.push(issue('categoryId', 'category_unavailable'));
      return undefined;
    }
    return String(candidate.value.value);
  }

  private interpretationOf(
    field: InterpretationFieldResult | undefined,
    key: DraftFieldKey,
    issues: MovementEntryDraftIssue[],
    required: boolean,
  ): InterpretationFieldCandidate | undefined {
    if (!field) {
      if (required) {
        issues.push(issue(key, 'field_missing'));
      }
      return undefined;
    }
    switch (field.interpretation.kind) {
      case 'resolved':
        return field.interpretation.candidate;
      case 'ambiguous':
        issues.push(issue(key, 'field_ambiguous'));
        return undefined;
      case 'missing':
        if (required) {
          issues.push(issue(key, 'field_missing'));
        }
        return undefined;
    }
  }
}

function issue(field: DraftFieldKey, code: MovementEntryDraftIssueCode): MovementEntryDraftIssue {
  return { field: fieldToDraft(field), code };
}

function fieldToDraft(field: DraftFieldKey): MovementEntryDraftField {
  return field;
}

function toDraftField(fieldKey: string): MovementEntryDraftField | undefined {
  return isDraftField(fieldKey) ? fieldKey : undefined;
}

function isDraftField(fieldKey: string): fieldKey is DraftFieldKey {
  return draftFields.has(fieldKey as DraftFieldKey);
}

File: core/src/main/kotlin/com/gonezo/application/orchestration/interpretation/InterpretationResultToMovementEntryDraftMapper.kt
package com.gonezo.application.orchestration.interpretation

import dev.solidcoder.interpretation.domain.FieldInterpretation
import dev.solidcoder.interpretation.domain.FieldKey
import dev.solidcoder.interpretation.domain.InterpretationResult
import dev.solidcoder.interpretation.domain.StructuredValue
import java.math.BigDecimal
import java.time.LocalDate

class InterpretationResultToMovementEntryDraftMapper(
  private val expectedSpecId: String = "movement-entry",
  private val expectedSpecVersion: String = "v1",
  private val minimumConfidence: Double = 0.5,
) {
  init {
    require(minimumConfidence in 0.0..1.0) { "minimum confidence must be between zero and one" }
  }

  fun map(result: InterpretationResult): MovementEntryDraft {
    require(result.specId.value == expectedSpecId) { "interpretation result spec id does not match the movement draft schema" }
    require(result.specVersion.value == expectedSpecVersion) { "interpretation result spec version does not match the movement draft schema" }

    val issues = mutableListOf<MovementEntryDraftIssue>()

    val type = mapType(result, issues)
    val amount = mapDecimal(result, "amount", MovementEntryDraftField.AMOUNT, issues)
    val occurredOn = mapDate(result, "occurredOn", MovementEntryDraftField.OCCURRED_ON, issues)
    val note = mapText(result, "note", MovementEntryDraftField.NOTE, issues)
    val categoryId = mapCategory(result, issues)

    result.issues.forEach { issue ->
      issues += MovementEntryDraftIssue(
        field = issue.fieldKey?.let(::fieldFromKey),
        code = MovementEntryDraftIssueCode.OUTPUT_INCOMPATIBLE,
      )
    }

    return MovementEntryDraft(
      type = type,
      amount = amount,
      occurredOn = occurredOn,
      note = note,
      categoryId = categoryId,
      issues = issues,
    )
  }

  private fun mapType(result: InterpretationResult, issues: MutableList<MovementEntryDraftIssue>): MovementEntryDraftType? {
    val interpretation = interpretationFor(result, "type")
    return when (interpretation) {
      FieldInterpretation.Missing -> {
        issues += issue(MovementEntryDraftField.TYPE, MovementEntryDraftIssueCode.FIELD_MISSING)
        null
      }
      is FieldInterpretation.Ambiguous -> {
        issues += issue(MovementEntryDraftField.TYPE, MovementEntryDraftIssueCode.FIELD_AMBIGUOUS)
        null
      }
      is FieldInterpretation.Resolved -> {
        val candidate = interpretation.candidate
        if (candidate.confidence.value < minimumConfidence) {
          issues += issue(MovementEntryDraftField.TYPE, MovementEntryDraftIssueCode.CONFIDENCE_INSUFFICIENT)
          return null
        }
        val enumValue = candidate.value as? StructuredValue.Enum
          ?: throw IllegalArgumentException("type must be encoded as an enum")
        when (enumValue.stableValue) {
          "expense" -> MovementEntryDraftType.EXPENSE
          "income" -> MovementEntryDraftType.INCOME
          else -> throw IllegalArgumentException("type must be expense or income")
        }
      }
    }
  }

  private fun mapDecimal(
    result: InterpretationResult,
    key: String,
    field: MovementEntryDraftField,
    issues: MutableList<MovementEntryDraftIssue>,
  ): BigDecimal? {
    val interpretation = interpretationFor(result, key)
    return when (interpretation) {
      FieldInterpretation.Missing -> {
        issues += issue(field, MovementEntryDraftIssueCode.FIELD_MISSING)
        null
      }
      is FieldInterpretation.Ambiguous -> {
        issues += issue(field, MovementEntryDraftIssueCode.FIELD_AMBIGUOUS)
        null
      }
      is FieldInterpretation.Resolved -> {
        val candidate = interpretation.candidate
        if (candidate.confidence.value < minimumConfidence) {
          issues += issue(field, MovementEntryDraftIssueCode.CONFIDENCE_INSUFFICIENT)
          return null
        }
        val value = candidate.value as? StructuredValue.Decimal
          ?: throw IllegalArgumentException("$key must be encoded as a decimal")
        if (value.value <= BigDecimal.ZERO) {
          throw IllegalArgumentException("amount must be greater than zero")
        }
        value.value
      }
    }
  }

  private fun mapDate(
    result: InterpretationResult,
    key: String,
    field: MovementEntryDraftField,
    issues: MutableList<MovementEntryDraftIssue>,
  ): LocalDate? {
    val interpretation = interpretationFor(result, key)
    return when (interpretation) {
      FieldInterpretation.Missing -> {
        issues += issue(field, MovementEntryDraftIssueCode.FIELD_MISSING)
        null
      }
      is FieldInterpretation.Ambiguous -> {
        issues += issue(field, MovementEntryDraftIssueCode.FIELD_AMBIGUOUS)
        null
      }
      is FieldInterpretation.Resolved -> {
        val candidate = interpretation.candidate
        if (candidate.confidence.value < minimumConfidence) {
          issues += issue(field, MovementEntryDraftIssueCode.CONFIDENCE_INSUFFICIENT)
          return null
        }
        val value = candidate.value as? StructuredValue.Date
          ?: throw IllegalArgumentException("$key must be encoded as a date")
        value.value
      }
    }
  }

  private fun mapText(
    result: InterpretationResult,
    key: String,
    field: MovementEntryDraftField,
    issues: MutableList<MovementEntryDraftIssue>,
  ): String? {
    val interpretation = interpretationFor(result, key)
    return when (interpretation) {
      FieldInterpretation.Missing -> {
        issues += issue(field, MovementEntryDraftIssueCode.FIELD_MISSING)
        null
      }
      is FieldInterpretation.Ambiguous -> {
        issues += issue(field, MovementEntryDraftIssueCode.FIELD_AMBIGUOUS)
        null
      }
      is FieldInterpretation.Resolved -> {
        val candidate = interpretation.candidate
        if (candidate.confidence.value < minimumConfidence) {
          issues += issue(field, MovementEntryDraftIssueCode.CONFIDENCE_INSUFFICIENT)
          return null
        }
        val value = when (val structuredValue = candidate.value) {
          is StructuredValue.Text -> structuredValue.value.trim()
          else -> throw IllegalArgumentException("$key must be encoded as text")
        }
        if (value.isBlank()) {
          throw IllegalArgumentException("$key cannot be blank")
        }
        value
      }
    }
  }

  private fun mapCategory(
    result: InterpretationResult,
    issues: MutableList<MovementEntryDraftIssue>,
  ): String? {
    val categoryField = result.fields.firstOrNull { it.key == FieldKey.of("categoryId") } ?: return null
    val interpretation = categoryField.interpretation
    return when (interpretation) {
      FieldInterpretation.Missing -> {
        issues += issue(MovementEntryDraftField.CATEGORY_ID, MovementEntryDraftIssueCode.FIELD_MISSING)
        null
      }
      is FieldInterpretation.Ambiguous -> {
        issues += issue(MovementEntryDraftField.CATEGORY_ID, MovementEntryDraftIssueCode.FIELD_AMBIGUOUS)
        null
      }
      is FieldInterpretation.Resolved -> {
        val candidate = interpretation.candidate
        if (candidate.confidence.value < minimumConfidence) {
          issues += issue(MovementEntryDraftField.CATEGORY_ID, MovementEntryDraftIssueCode.CONFIDENCE_INSUFFICIENT)
          return null
        }
        val value = candidate.value as? StructuredValue.Enum
          ?: throw IllegalArgumentException("categoryId must be encoded as an enum")
        value.stableValue
      }
    }
  }

  private fun interpretationFor(result: InterpretationResult, key: String): FieldInterpretation = result.fields
    .firstOrNull { it.key == FieldKey.of(key) }
    ?.interpretation
    ?: throw IllegalArgumentException("interpretation result is missing field $key")

  private fun fieldFromKey(key: dev.solidcoder.interpretation.domain.FieldKey): MovementEntryDraftField? = when (key.value) {
    "type" -> MovementEntryDraftField.TYPE
    "amount" -> MovementEntryDraftField.AMOUNT
    "occurredOn" -> MovementEntryDraftField.OCCURRED_ON
    "note" -> MovementEntryDraftField.NOTE
    "categoryId" -> MovementEntryDraftField.CATEGORY_ID
    else -> null
  }

  private fun issue(field: MovementEntryDraftField, code: MovementEntryDraftIssueCode) = MovementEntryDraftIssue(field, code)
}
